File: src/modules/auth/api/socialApi.ts
import type { AuthApi } from './authApi'

export type UserMode = 'PUBLIC' | 'PRIVATE'

export type UserProfile = {
  id: string
  email: string
  displayName: string
  mode: UserMode
  appearInSearches: boolean
  createdAt: string
  updatedAt: string
}

export type PublicUser = UserProfile

export type ChatRequestStatus = 'PENDING' | 'ACCEPTED' | 'REJECTED'

export type ChatRequest = {
  id: string
  fromUserId: string
  toUserId: string
  fromUser: UserProfile
  toUser: UserProfile
  status: ChatRequestStatus
  createdAt: string
  updatedAt: string
}

export type SavedChat = {
  id: string
  chatRequestId: string
  savedByUserId: string
  chatRequest: ChatRequest
  savedBy: UserProfile
  savedAt: string
}

export type BlockedUser = UserProfile

export type UpdateProfilePayload = {
  displayName: string
}

export type RespondToChatRequestResponse = {
  request: ChatRequest
  roomId: string | null
}

export const createSocialApi = (authApi: AuthApi) => {
  const getMyProfile = () => authApi.request<UserProfile>('/users/me')

  const updateMyProfile = (payload: UpdateProfilePayload) =>
    authApi.request<UserProfile>('/users/me', {
      method: 'PATCH',
      body: JSON.stringify({ displayName: payload.displayName }),
    })

  const updateMyMode = (mode: UserMode) =>
    authApi.request<UserProfile>('/users/me/mode', {
      method: 'PATCH',
      body: JSON.stringify({ mode }),
    })

  const deleteMyAccount = () =>
    authApi.request<void>('/users/me', {
      method: 'DELETE',
    })

  const listPublicUsers = () => authApi.request<PublicUser[]>('/users/public')

  const searchPublicUsers = (query: string) =>
    authApi.request<PublicUser[]>(`/users/search?q=${encodeURIComponent(query.trim())}`)

  const listSentChatRequests = () => authApi.request<ChatRequest[]>('/chat/requests/sent')

  const listReceivedChatRequests = () => authApi.request<ChatRequest[]>('/chat/requests/received')

  const sendChatRequest = (toDisplayName: string) =>
    authApi.request<ChatRequest>('/chat/requests', {
      method: 'POST',
      body: JSON.stringify({ toDisplayName }),
    })

  const respondToChatRequest = (chatRequestId: string, accept: boolean) =>
    authApi.request<RespondToChatRequestResponse>(`/chat/requests/${chatRequestId}/respond`, {
      method: 'POST',
      body: JSON.stringify({ accept }),
    })

  const cancelChatRequest = (chatRequestId: string) =>
    authApi.request<void>(`/chat/requests/${chatRequestId}`, {
      method: 'DELETE',
    })

  const listSavedChats = () => authApi.request<SavedChat[]>('/chat/saved')

  const saveChat = (chatRequestId: string) =>
    authApi.request<SavedChat>(`/chat/requests/${chatRequestId}/save`, {
      method: 'POST',
    })

  const deleteSavedChat = (savedChatId: string) =>
    authApi.request<void>(`/chat/saved/${savedChatId}`, {
      method: 'DELETE',
    })

  const listBlockedUsers = () => authApi.request<BlockedUser[]>('/users/me/blocked')

  const blockUser = (blockedUserId: string) =>
    authApi.request<void>(`/users/${blockedUserId}/block`, {
      method: 'POST',
    })

  const unblockUser = (blockedUserId: string) =>
    authApi.request<void>(`/users/${blockedUserId}/block`, {
      method: 'DELETE',
    })

  const updateAppearInSearches = (appearInSearches: boolean) =>
    authApi.request<UserProfile>('/users/me/appear-in-searches', {
      method: 'PATCH',
      body: JSON.stringify({ appearInSearches }),
    })

  return {
    getMyProfile,
    updateMyProfile,
    updateMyMode,
    deleteMyAccount,
    listPublicUsers,
    searchPublicUsers,
    listSentChatRequests,
    listReceivedChatRequests,
    sendChatRequest,
    respondToChatRequest,
    cancelChatRequest,
    listSavedChats,
    saveChat,
    deleteSavedChat,
    listBlockedUsers,
    blockUser,
    unblockUser,
    updateAppearInSearches,
  }
}

export type SocialApi = ReturnType<typeof createSocialApi>

File: flutter/lib/features/home/data/social_api.dart
import '../../../core/network/api_client.dart';
import '../../auth/data/token_store.dart';
import '../domain/home_models.dart';

/// Social API for managing user profiles, chat requests, saved chats, and blocked users
/// Maps to the React app's socialApi
class SocialApi {
  SocialApi({required ApiClient client, required TokenStore tokenStore})
      : _client = client,
        _tokenStore = tokenStore;

  final ApiClient _client;
  final TokenStore _tokenStore;

  Future<Map<String, String>> _authHeaders() async {
    final String? token = await _tokenStore.readToken();
    if (token == null || token.isEmpty) {
      throw Exception('No access token available');
    }
    return <String, String>{
      'Authorization': 'Bearer $token',
    };
  }

  // User Profile Management

  Future<UserProfile> getMyProfile() async {
    final Map<String, dynamic> response = await _client.getJson(
      '/users/me',
      headers: await _authHeaders(),
    );
    return UserProfile.fromJson(response);
  }

  Future<UserProfile> updateMyProfile({required String displayName}) async {
    final Map<String, dynamic> response = await _client.patchJson(
      '/users/me',
      body: <String, dynamic>{'displayName': displayName.trim()},
      headers: await _authHeaders(),
    );
    return UserProfile.fromJson(response);
  }

  Future<UserProfile> updateMyMode({required UserMode mode}) async {
    final Map<String, dynamic> response = await _client.patchJson(
      '/users/me/mode',
      body: <String, dynamic>{'mode': mode.value},
      headers: await _authHeaders(),
    );
    return UserProfile.fromJson(response);
  }

  Future<UserProfile> updateAppearInSearches({required bool appearInSearches}) async {
    final Map<String, dynamic> response = await _client.patchJson(
      '/users/me/appear-in-searches',
      body: <String, dynamic>{'appearInSearches': appearInSearches},
      headers: await _authHeaders(),
    );
    return UserProfile.fromJson(response);
  }

  Future<void> deleteMyAccount() async {
    await _client.deleteJson(
      '/users/me',
      headers: await _authHeaders(),
    );
  }

  // User Search and Discovery

  Future<List<UserProfile>> listPublicUsers() async {
    final dynamic response = await _client.get(
      '/users/public',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <UserProfile>[];
    }

    return response
        .map((dynamic item) => UserProfile.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<List<UserProfile>> searchPublicUsers(String query) async {
    final dynamic response = await _client.get(
      '/users/search?q=${Uri.encodeQueryComponent(query.trim())}',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <UserProfile>[];
    }

    return response
        .map((dynamic item) => UserProfile.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  // Chat Request Management

  Future<List<ChatRequest>> listSentChatRequests() async {
    final dynamic response = await _client.get(
      '/chat/requests/sent',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <ChatRequest>[];
    }

    return response
        .map((dynamic item) => ChatRequest.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<List<ChatRequest>> listReceivedChatRequests() async {
    final dynamic response = await _client.get(
      '/chat/requests/received',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <ChatRequest>[];
    }

    return response
        .map((dynamic item) => ChatRequest.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<void> sendChatRequest({required String toDisplayName}) async {
    await _client.postJson(
      '/chat/requests',
      body: <String, dynamic>{'toDisplayName': toDisplayName.trim()},
      headers: await _authHeaders(),
    );
    // Note: POST response doesn't include fromUser/toUser objects
    // Caller should reload dashboard data to get complete request info
  }

  Future<RespondToChatRequestResponse> respondToChatRequest({
    required String chatRequestId,
    required bool accept,
  }) async {
    final Map<String, dynamic> response = await _client.postJson(
      '/chat/requests/$chatRequestId/respond',
      body: <String, dynamic>{'accept': accept},
      headers: await _authHeaders(),
    );
    return RespondToChatRequestResponse.fromJson(response);
  }

  Future<void> cancelChatRequest({required String chatRequestId}) async {
    await _client.deleteJson(
      '/chat/requests/$chatRequestId',
      headers: await _authHeaders(),
    );
  }

  Future<void> removeRecentChat({required String chatRequestId}) async {
    await _client.deleteJson(
      '/chat/requests/$chatRequestId/remove',
      headers: await _authHeaders(),
    );
  }

  // Saved Chats Management

  Future<List<SavedChat>> listSavedChats() async {
    final dynamic response = await _client.get(
      '/chat/saved',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <SavedChat>[];
    }

    return response
        .map((dynamic item) => SavedChat.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<void> saveChat({required String chatRequestId}) async {
    await _client.postJson(
      '/chat/requests/$chatRequestId/save',
      headers: await _authHeaders(),
    );
  }

  Future<void> deleteSavedChat({required String savedChatId}) async {
    await _client.deleteJson(
      '/chat/saved/$savedChatId',
      headers: await _authHeaders(),
    );
  }

  // Blocked Users Management

  Future<List<UserProfile>> listBlockedUsers() async {
    final dynamic response = await _client.get(
      '/users/me/blocked',
      headers: await _authHeaders(),
    );

    if (response is! List) {
      return <UserProfile>[];
    }

    return response
        .map((dynamic item) => UserProfile.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<void> blockUser({required String blockedUserId}) async {
    await _client.postJson(
      '/users/$blockedUserId/block',
      headers: await _authHeaders(),
    );
  }

  Future<void> unblockUser({required String blockedUserId}) async {
    await _client.deleteJson(
      '/users/$blockedUserId/block',
      headers: await _authHeaders(),
    );
  }
}
